tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,262 @@ def option_price(*,
+          name="continuous_dividends")
+
+    spots = tf.convert_to_tensor(
+        spots, dtype=dtype, name="spots")
+    strikes = tf.convert_to_tensor(
+        strikes, dtype=dtype, name="strikes")
+    barriers = tf.convert_to_tensor(
+        barriers, dtype=dtype, name="barriers")
+    rebates = tf.convert_to_tensor(
+        rebates, dtype=dtype, name="rebates")
+    volatilities = tf.convert_to_tensor(
+        volatilities, dtype=dtype, name="volatilities")
+    expiries = tf.convert_to_tensor(
+        expiries, dtype=dtype, name="expiries")
+
+    is_barrier_down = tf.convert_to_tensor(is_barrier_down, dtype=tf.int32,
+                                           name="is_barrier_down")
+    is_knock_out = tf.convert_to_tensor(is_knock_out, dtype=tf.int32,
+                                        name="is_knock_out")
+    is_call_options = tf.convert_to_tensor(is_call_options, dtype=tf.int32,
+                                           name="is_call_options")
+    indices = tf.bitwise.left_shift(
+        is_barrier_down, 2) + tf.bitwise.left_shift(
+            is_knock_out, 1) + is_call_options
+
+    # Masks select the appropriate terms for integral approximations
+    # Interals are seperated by algebraic terms and probability
+    # distribution terms. This give 12 different terms per matrix
+    # (6 integrals, 2 terms each)
+    # shape = [1, 12]
+    mask_matrix_greater_strike = tf.constant([
+        [1, 1, -1, -1, 0, 0, 1, 1, 1, 1, 0, 0], # up and in put
+        [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0], # up and in call
+        [0, 0, 1, 1, -1, -1, 1, 1, 0, 0, 1, 1], # down and in put
+        [0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0], # down and in call
+        [0, 0, 1, 1, 0, 0, -1, -1, 0, 0, 1, 1], # up and out put
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1], # up and out call
+        [1, 1, -1, -1, 1, 1, -1, -1, 0, 0, 1, 1], # down and out put
+        [1, 1, 0, 0, -1, -1, 0, 0, 0, 0, 1, 1]]) # down and out call
+
+    mask_matrix_lower_strike = tf.constant([
+        [0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0], # up and in put
+        [0, 0, 1, 1, -1, -1, 1, 1, 1, 1, 0, 0], # up and in call
+        [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0], # down and in put
+        [1, 1, -1, -1, 0, 0, 1, 1, 1, 1, 0, 0], # down and in call
+        [1, 1, 0, 0, -1, -1, 0, 0, 0, 0, 1, 1], # up and out put
+        [1, 1, -1, -1, 1, 1, -1, -1, 0, 0, 1, 1], # up and out call
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1], # down and out put
+        [0, 0, 1, 1, 0, 0, -1, -1, 0, 0, 1, 1]]) # down and out call
+
+    # Get mask for each option
+    masks_lower = tf.gather(mask_matrix_lower_strike, indices)
+    masks_greater = tf.gather(mask_matrix_greater_strike, indices)
+    strikes_greater = tf.expand_dims(strikes > barriers, axis=-1)
+    masks = tf.where(strikes_greater, masks_greater, masks_lower)
+    masks = tf.cast(masks, dtype=dtype)
+
+    call_or_put = tf.where(tf.equal(is_call_options, 0), -1., 1.)
```

Comment:
`call_or_put` always has float32 dtype. How about 
`one = tf.constant(1, dtype=dtype)`
and then
`call_or_put = tf.where(tf.equal(is_call_options, 0), -one, one)`

Reply:
Done!